federation-js/src/composition/DirectiveMetadata.ts

```diff
@@ -192,3 +179,23 @@ export class DirectiveMetadata {
     }
   }
 }
+
+function mergeDirectiveUsages(
+  destination: DirectiveUsages | undefined,
+  source: DirectiveUsages,
+): DirectiveUsages {
+  let directiveUsages = destination;
+  if (!directiveUsages || directiveUsages.size === 0) {
+    directiveUsages = source;
+  } else {
+    for (const [directiveName, usages] of source.entries()) {
+      const existingUsages = directiveUsages.get(directiveName);
+      if (existingUsages && existingUsages.length > 0) {
+        existingUsages.push(...usages);
+      } else {
+        directiveUsages.set(directiveName, usages);
```

Comment:
This is moving the exact Array reference `usages` that lives in `source` into `destination`, and then maybe another iteration (or another call?) that runs the line that is two above this mutates the array.  So you might end up surprisingly mutating `source`. Maybe that is fine in this context but maybe it could be avoided?

Perhaps the utility function mapGetOrSet that we added a bit ago would help.  `mapGetOrSet(directiveUsages, directiveName, []).push(...usages)` can replace the entire body of the for loop, right?

Reply:
e6c6c12